File: shell/src/main/resources/org/efaps/js/classes/Table.js
/*
 * Copyright 2006 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:          tmo
 * Revision:        $Rev$
 * Last Changed:    $Date$
 * Last Changed By: $Author$
 */

importClass(Packages.org.efaps.db.Instance);
importClass(Packages.org.efaps.db.SearchQuery);

function Table(_name)  {
  this.object = new EFapsInstance("Admin_UI_Table", _name);
  this.instance = new Instance(Shell.getContext(), this.object.oid);

  // add helper methods for getter and setter 
  this._setAttrValue      = Abstract.prototype._setAttrValue;
  this._getAttrValue      = Abstract.prototype._getAttrValue;

  // add getter and setter methods
  this.getInstance        = Abstract.prototype.getInstance;
  this.getId              = Abstract.prototype.getId;
  this.getOid             = Abstract.prototype.getOid;
  this.getType            = Abstract.prototype.getType;
  this.getName            = Abstract.prototype.getName;
  this.getRevision        = Abstract.prototype.getRevision;
  this.setRevision        = Abstract.prototype.setRevision;

  // add field methods
  this._writeFields       = UICollection.prototype._writeFields;
  this.cleanupFields      = UICollection.prototype.cleanupFields;
  this.addField           = UICollection.prototype.addField;

  // add property methods
  this._writeProperties   = Abstract.prototype._writeProperties;
  this.addProperty        = Abstract.prototype.addProperty;
  this.cleanupProperties  = Abstract.prototype.cleanupProperties;
  this.deleteProperty     = Abstract.prototype.addProperty;
  this.printProperties    = Abstract.prototype.printProperties;

  // add common methods
  this._writeHeader       = Abstract.prototype._writeHeader;
  this.cleanup            = UICollection.prototype.cleanup;
  this.writeUpdateScript  = UICollection.prototype.writeUpdateScript;
}

///////////////////////////////////////////////////////////////////////////////
// instance variables

/**
 * Prefix of script file names.
 */
Table.prototype.FILE_PREFIX = new String("TABLE_");

/**
 * Name of the variable used in update scripts.
 */
Table.prototype.VARNAME     = new String("TABLE");

///////////////////////////////////////////////////////////////////////////////
// common methods

Table.prototype.update = function(_fileName, _objName)  {
  if (this.getOid()==null || this.getOid()=="" || this.getOid()=="0")  {
    print("  - create");
    this.object.Name = _objName;
    this.object.create();
    this.instance = new Instance(Shell.getContext(), this.object.oid);
  } else  {
    print("  - cleanup");
    this.cleanup();
  }
  TABLE = this;
  print("  - import");
  load(_fileName);
  TABLE = null;
}

///////////////////////////////////////////////////////////////////////////////
// static functions

/**
 * Extract the table name from a file name.
 *
 * @param (String)_fileName   file name
 * @return extracted table name
 */
function getTableNameFromFileName(_fileName)  {
  var tmp = new String(_fileName);
  return tmp.substring(Table.prototype.FILE_PREFIX.length, tmp.length-3);
}

/**
 * Import exact one table.
 *
 * @param (File)_fileName   file name of the table to import
 */
function importTable(_fileName)  {
  var fileName = new File(_fileName);
  if (fileName.getName().startsWith(Table.prototype.FILE_PREFIX) && fileName.getName().endsWith(".js"))  {
    var objName = getTableNameFromFileName(fileName.getName());
    print("Import Table '"+objName+"'");
    var imp = new Table(objName);
    imp.update(_fileName, objName);
  }
}

/**
 * A list of Table Definition files is imported. The algorithmen checks first
 * for existance of a table and creates the not existant tables.
 *
 * @param _fileList array with list of files
 */
function importTables(_fileList)  {
  print("");
  print("Import Tables");
  print("~~~~~~~~~~~~~");
  for (indx in _fileList)  {
    importTable(_fileList[indx]);
  }
  print("");
}


/*
  var codeRev = getCodeRevision(path+"/"+fileName);
  var dbRev = TABLE.getRevision();

  if (codeRev==null || codeRev.length==0 || codeRev!=dbRev)  {
    if (codeRev!=null)  {
      print("  - update to revision '"+codeRev+"'");
      TABLE.setRevision(codeRev);
    } else  {
      print("  - update always (no code revision)");
    }
*/

/**
 * @param _path   (String)  files are created in this path
 * @param _match  (String)  match for menu names
 * @param _author (String)  author used in the header
 */
function createScriptTables(_path, _match, _author)  {
  var query = new SearchQuery();
  query.setQueryTypes(Shell.getContext(), "Admin_UI_Table");
  query.addWhereExprEqValue(Shell.getContext(), "Name", _match);
  query.addSelect(Shell.getContext(), "Name");
  
  query.execute(Shell.getContext());
  
  print("");
  print("Create Table Scripts:");
  print("~~~~~~~~~~~~~~~~~~~~~");
  while (query.next())  {
    var name = query.get(Shell.getContext(), "Name");
    print("  - '"+  name  +"'");
    var table = new Table(name);
    table.writeUpdateScript(_path, _author);
  }

  query.close();
}

File: src/main/java/org/efaps/admin/ui/Table.java
/*
 * Copyright 2003 - 2016 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.efaps.admin.ui;

import java.util.UUID;

import org.efaps.ci.CIAdminUserInterface;
import org.efaps.util.cache.CacheReloadException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author The eFaps Team
 * TODO:
 *          description
 */
public class Table
    extends AbstractCollection
    implements Cloneable
{

    /**
     * Logging instance used in this class.
     */
    protected static final Logger LOG = LoggerFactory.getLogger(Table.class);

    /**
     * Needed for serialization.
     */
    private static final long serialVersionUID = 1L;

    /**
     * This is the constructor to set the id and the name.
     *
     * @param _id id of the new table
     * @param _uuid UUID of the new table
     * @param _name name of the new table
     */
    public Table(final Long _id,
                 final String _uuid,
                 final String _name)
    {
        super(_id, _uuid, _name);
    }

    /**
     * Creates and returns a copy of this table object.
     *
     * @return cloned table
     */
    public Table cloneTable()
    {
        Table ret = null;
        try {
            ret = (Table) super.clone();
        } catch (final CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return ret;
    }

    /**
     * Returns for given parameter <i>_id</i> the instance of class
     * {@link Table}.
     *
     * @param _id id to search in the cache
     * @return instance of class {@link Table}
     * @throws CacheReloadException on error
     */
    public static Table get(final long _id)
        throws CacheReloadException
    {
        return AbstractUserInterfaceObject.<Table>get(_id, Table.class, CIAdminUserInterface.Table.getType());
    }

    /**
     * Returns for given parameter <i>_name</i> the instance of class
     * {@link Table}.
     *
     * @param _name name to search in the cache
     * @return instance of class {@link Table}
     * @throws CacheReloadException on error
     */
    public static Table get(final String _name)
        throws CacheReloadException
    {
        return AbstractUserInterfaceObject.<Table>get(_name, Table.class, CIAdminUserInterface.Table.getType());
    }

    /**
     * Returns for given parameter <i>UUID</i> the instance of class
     * {@link Table}.
     *
     * @param _uuid UUID to search in the cache
     * @return instance of class {@link Table}
     * @throws CacheReloadException on error
     */
    public static Table get(final UUID _uuid)
        throws CacheReloadException
    {
        return AbstractUserInterfaceObject.<Table>get(_uuid, Table.class, CIAdminUserInterface.Table.getType());
    }
}
